lib/utils/fanoutbuffer/buffer.go

```diff
@@ -0,0 +1,406 @@
+	// longer waits before unhealthy cursors are ejected.
+	GracePeriod time.Duration
+
+	// Clock is used to override default time-behaviors in tests.
+	Clock clockwork.Clock
+}
+
+// SetDefaults sets default config parameters.
+func (c *Config) SetDefaults() {
+	if c.Capacity == 0 {
+		c.Capacity = 512
+	}
+
+	if c.GracePeriod == 0 {
+		c.GracePeriod = time.Second * 30
+	}
+
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+}
+
+// Buffer is a circular buffer that keeps track of how many cursors exist, and how many have seen
+// each item, so that it knows when items can be cleared. If one or more cursors fall behind, the items
+// they have yet to see go into a temporary backlog of infinite size. If the backlog persists for greater
+// than the allowed grace period, it is cleared and all cursors still in the backlog fail on their next
+// attempt to read.
+type Buffer[T any] struct {
+	rw sync.RWMutex
+
+	// cfg holds buffer configuration.
+	cfg Config
+
+	// notify is closed and replaced each time an item is added to wake any cursors
+	// that are fully caught up.
+	notify chan struct{}
+
+	// closed indicates that the buffer has been closed.
+	closed bool
+
+	// cursors is the total number of cursors that exist. used to determine the number of cursors expected
+	// to observe each item.
+	cursors uint64
+
+	// pos is the absolute position of the oldest elemet in the ring, meaning that pos%Capacity is the index
+	// of that element.
+	pos uint64
+
+	// len is the total number of elements in the ring (i.e. the index of the newest element is at `(pos + len - 1) % Capacity`).
+	len uint64
+
+	// ring is the fixed-size circular buffer that serves as our primary storage location.
+	ring []entry[T]
+
+	// overflowStart is the time at which we began needing to overflow items.
+	overflowStart time.Time
+
+	// overflow is a variable length buffer of items that still need to be observed.
+	overflow []entry[T]
+}
+
+// entry represents an item in a buffer
+type entry[T any] struct {
+	item T
+	// wait is the number of cursors that still need to observe this item. each cursor
+	// decrements this value when it observes the item. items with wait==0 are cleaned
+	// up. Note that this atomic is only safe to decrement while read lock is held. Entries
+	// may be moved while write lock is held.
+	wait atomic.Uint64
+}
+
+// NewBuffer creates a new fanout buffer instance.
+func NewBuffer[T any](cfg Config) *Buffer[T] {
+	cfg.SetDefaults()
+	return &Buffer[T]{
+		cfg:    cfg,
+		notify: make(chan struct{}),
+		ring:   make([]entry[T], cfg.Capacity),
+	}
+}
+
+// Close permanently closes the fanout buffer. Note that all cursors are termianted immediately
+// and may therefore miss items that had not been read at the time Close is called.
+func (b *Buffer[T]) Close() {
+	var notify chan struct{}
+	b.write(func() {
+		notify = b.notify
+		b.notify = make(chan struct{})
```

Comment:
Leftover from an earlier iteration where `Buffer` supported `Reset` instead of close.  Ended up complicating concurrency a lot, so I moved the reset up to the `Fanout` level and just closed buffers and created new ones on reset.  Will remove.